### PR TITLE
exp: Drop `checkpoints`

### LINKED ISSUE #9221: Drop/Revisit usage of `checkpoints`
> Are we in agreement that we don't support checkpoint anymore? (I'm personally still not convinced. Primarily because I'm not sure we have a decent replacement for this. I think it's needed if we were to remove this).

_Originally posted by @shcheklein in https://github.com/iterative/dvc.org/issues/4415#issuecomment-1476721592_

---

I don't really think we need a built-in replacement/solution in DVC to handle the checkpoints use case (which I am still unsure what that is). 

People should handle interruption and resuming through the ML framework and DVC already provides convenient tools to wrap it (params, `persist`, run-cache). 

My main points about dropping checkpoints are:

- The current solution doesn't provide value while coming at an important cost of code/docs maintenance. 

- Induces users into incorrect workflows and the required changes in the code are not properly explained anywhere.

- Introduces ambiguous/unexpected behavior when executing more complex/realistic pipelines (i.e. how are downstream stages after `checkpoint: true` supposed to be executed?; what about `foreach` or `dvc.yaml` with more than 1 model?)

---

As an example of the second point, here are the things that are "incorrect" in this repo (same applies to the example in https://dvc.org/doc/user-guide/experiment-management/checkpoints):

- Optimizer state is not handled.

The `state_dict` of the optimizer should be also considered when loading/saving.

- No learning rate scheduler.

I would dare to say that using a fixed learning rate would never result in a better model than using _any_ kind of lr scheduler. 

It would also require to be handled when loading/saving (which connects with the issues in the next point).

- Epochs are being handled (arguably) incorrectly.

When picking a checkpoint and resuming from it, the `epochs` param is now treated as `epochs + epochs_completed_at_checkpoint` which differs with the meaning when training without resuming where `epochs` params reflects the total number.

- After resuming from a checkpoint, the experiments can't be reproduced easily.

Let's say we have a completed experiments that was using checkpoints:

```yaml
# EXP_A
lr: 0.003
weight_decay: 0
epochs: 15
```

If I run:

```console
$ dvc exp apply EXP_A_CHECKPOINT_5
$ dvc exp run -S lr=99 weight_decay=0 epochs=10
```

It is not possible to reproduce the experiment with a single command. We would have to run the exact combination of `exp apply` and `exp run`.
It is not possible to reproduce the experiment at all if the checkpoints are deleted.

- Resumed experiments are not differentiable after persisting.

Let's say I have another experiment completed using checkpoints:

```yaml
# EXP_B
lr: 0.1
weight_decay: 10
epochs: 40
```

And I run:

```console
$ dvc exp apply EXP_B_CHECKPOINT_39
$ dvc exp run -S lr=99 weight_decay=0 epochs=10
```

Persisting this experiment or the one from the previous point will result in an equivalent state in the repo regarding `params` and `step` metric, even though the training that led to the resulting model is completely different.

### PR DESCRIPTION
**Pull request by daavoo**
Closes #9221
Closes https://github.com/iterative/dvc/issues/8750
Closes https://github.com/iterative/dvc/issues/8622
Closes https://github.com/iterative/dvc/issues/8009
Closes https://github.com/iterative/dvc/issues/7381
Closes https://github.com/iterative/dvc/issues/6470
Closes https://github.com/iterative/dvc/issues/6194

As part of 3.0 release - (see checklist: #7093)

- Docs: https://github.com/iterative/dvc.org/pull/4442

### CODE REVIEW

**Comment by shcheklein** [comment]
That was fast :) I don't think we discussed to drop this in the DVC itself yet. @dberenbaum wdyt? are we good with this? or we want to hide it first and think about the alternative?

**Comment by daavoo** [comment]
> That was fast :) I don't think we discussed to drop this in the DVC itself yet. @dberenbaum wdyt? are we good with this? or we want to hide it first and think about the alternative?

To clarify. It is not my intention to merge and drop it right away. I am planning to discuss with the team a branching strategy for the major release and that is where these draft P.Rs would go.

I just wanted to have the work done so:
- It is easier to just rebase on this branch
- When we discuss the major release, it doesn't bias the fact that "we still have to do this and that takes time", because it is done.

**Review by pmrowla — APPROVED** [review]
